File: scripts/precise-calculation.js
const sqlite3 = require('sqlite3').verbose();
const path = require('path');
const fs = require('fs');

const dbPath = path.join(__dirname, '..', 'assets', 'ecdict-sqlite-28', 'stardict.db');
const db = new sqlite3.Database(dbPath, sqlite3.OPEN_READONLY);

console.log('=== 精确计算 30MB 可容纳词条数 ===\n');

const TARGET_SIZE_MB = 30;
const TARGET_SIZE_BYTES = TARGET_SIZE_MB * 1024 * 1024;

console.log(`目标大小: ${TARGET_SIZE_MB} MB = ${TARGET_SIZE_BYTES.toLocaleString()} 字节\n`);

// 创建临时测试数据库来精确测量
const tempDbPath = path.join(__dirname, '..', 'temp', 'test.db');

// 确保temp目录存在
if (!fs.existsSync(path.join(__dirname, '..', 'temp'))) {
  fs.mkdirSync(path.join(__dirname, '..', 'temp'));
}

// 测试不同词条数的实际大小
const testSizes = [1000, 5000, 10000, 20000, 30000, 50000];
let currentTest = 0;

console.log('正在创建测试数据库，精确测量实际大小...\n');

function testNextSize() {
  if (currentTest >= testSizes.length) {
    analyzeResults();
    return;
  }
  
  const testSize = testSizes[currentTest];
  console.log(`测试 ${testSize.toLocaleString()} 词条...`);
  
  // 删除旧的测试数据库
  if (fs.existsSync(tempDbPath)) {
    fs.unlinkSync(tempDbPath);
  }
  
  // 创建新的测试数据库
  const testDb = new sqlite3.Database(tempDbPath);
  
  testDb.serialize(() => {
    // 创建表结构（只有主键索引）
    testDb.run(`
      CREATE TABLE dict (
        word TEXT PRIMARY KEY COLLATE NOCASE,
        phonetic TEXT,
        translation TEXT,
        collins INTEGER DEFAULT 0,
        oxford INTEGER DEFAULT 0
      )
    `);
    
    // 插入数据
    const stmt = testDb.prepare(`
      INSERT INTO dict (word, phonetic, translation, collins, oxford)
      VALUES (?, ?, ?, ?, ?)
    `);
    
    db.all(`
      SELECT word, phonetic, translation, collins, oxford
      FROM stardict
      WHERE oxford > 0 OR collins > 0 OR (bnc > 0 AND bnc <= 50000)
      LIMIT ${testSize}
    `, (err, rows) => {
      if (err) {
        console.error('查询错误:', err);
        return;
      }
      
      rows.forEach(row => {
        // 压缩 translation（中度压缩）
        let translation = row.translation || '';
        translation = translation
          .replace(/\[网络\][^\n]*/g, '')
          .replace(/\[地名\][^\n]*/g, '')
          .replace(/\[人名\][^\n]*/g, '')
          .replace(/\s+/g, ' ')
          .trim()
          .substring(0, 150);
        
        stmt.run(
          row.word,
          row.phonetic,
          translation,
          row.collins || 0,
          row.oxford || 0
        );
      });
      
      stmt.finalize(() => {
        // VACUUM 压缩数据库
        testDb.run('VACUUM', (err) => {
          if (err) {
            console.error('VACUUM错误:', err);
            return;
          }
          
          testDb.close(() => {
            // 测量文件大小
            const stats = fs.statSync(tempDbPath);
            const sizeMB = (stats.size / 1024 / 1024).toFixed(2);
            const bytesPerWord = (stats.size / testSize).toFixed(2);
            const percentage = ((stats.size / TARGET_SIZE_BYTES) * 100).toFixed(1);
            
            console.log(`  实际大小: ${sizeMB} MB (${stats.size.toLocaleString()} 字节)`);
            console.log(`  每词占用: ${bytesPerWord} 字节`);
            console.log(`  占用率: ${percentage}%\n`);
            
            results.push({
              words: testSize,
              bytes: stats.size,
              sizeMB: parseFloat(sizeMB),
              bytesPerWord: parseFloat(bytesPerWord),
              percentage: parseFloat(percentage)
            });
            
            currentTest++;
            testNextSize();
          });
        });
      });
    });
  });
}

const results = [];
testNextSize();

function analyzeResults() {
  console.log('=== 实际测试结果 ===\n');
  console.log('词条数'.padEnd(12) + 
              '数据库大小'.padStart(15) + 
              '每词占用'.padStart(12) + 
              '占用率'.padStart(10));
  console.log('-'.repeat(50));
  
  results.forEach(r => {
    console.log(
      r.words.toLocaleString().padEnd(12) +
      (r.sizeMB + ' MB').padStart(15) +
      (r.bytesPerWord + ' B').padStart(12) +
      (r.percentage + '%').padStart(10)
    );
  });
  
  console.log('-'.repeat(50));
  console.log(`目标限制: ${TARGET_SIZE_MB} MB\n`);
  
  // 线性回归计算最大词条数
  if (results.length >= 2) {
    // 使用最后两个数据点计算平均每词占用
    const lastResult = results[results.length - 1];
    const avgBytesPerWord = lastResult.bytesPerWord;
    
    console.log('=== 容量计算 ===\n');
    console.log(`基于 ${lastResult.words.toLocaleString()} 词的测试：`);
    console.log(`平均每词占用: ${avgBytesPerWord} 字节\n`);
    
    // 计算不同压缩程度
    const compressionLevels = [
      { name: '当前压缩(150字符)', factor: 1.0 },
      { name: '轻度压缩(200字符)', factor: 1.15 },
      { name: '重度压缩(100字符)', factor: 0.85 },
      { name: '极限压缩(50字符)', factor: 0.70 },
    ];
    
    console.log('不同压缩策略下的最大词条数：\n');
    console.log('压缩策略'.padEnd(25) + '每词占用'.padStart(12) + '最大词条数'.padStart(15) + '实际大小'.padStart(12));
    console.log('-'.repeat(65));
    
    compressionLevels.forEach(level => {
      const bytesPerWord = avgBytesPerWord * level.factor;
      const maxWords = Math.floor(TARGET_SIZE_BYTES / bytesPerWord);
      const actualSizeMB = (maxWords * bytesPerWord / 1024 / 1024).toFixed(2);
      
      console.log(
        level.name.padEnd(25) +
        (bytesPerWord.toFixed(1) + ' B').padStart(12) +
        maxWords.toLocaleString().padStart(15) +
        (actualSizeMB + ' MB').padStart(12)
      );
    });
    
    console.log('\n');
  }
  
  // 分析成分占比
  analyzeComponents();
}

function analyzeComponents() {
  console.log('=== 数据库成分分析 ===\n');
  
  // 使用 50000 词的测试数据库
  const result50k = results.find(r => r.words === 50000);
  
  if (!result50k) {
    console.log('没有 50000 词的测试数据\n');
    compareWithOthers();
    return;
  }
  
  console.log(`基于 50,000 词的数据库 (${result50k.sizeMB} MB)：\n`);
  
  // 重新打开测试数据库分析
  const testDb = new sqlite3.Database(tempDbPath, sqlite3.OPEN_READONLY);
  
  testDb.get(`
    SELECT 
      SUM(LENGTH(word)) as word_bytes,
      SUM(LENGTH(phonetic)) as phonetic_bytes,
      SUM(LENGTH(translation)) as translation_bytes,
      COUNT(*) * 8 as int_bytes,
      COUNT(*) as count
    FROM dict
  `, (err, row) => {
    if (err) {
      console.error('分析错误:', err);
      testDb.close();
      compareWithOthers();
      return;
    }
    
    const wordBytes = row.word_bytes || 0;
    const phoneticBytes = row.phonetic_bytes || 0;
    const translationBytes = row.translation_bytes || 0;
    const intBytes = row.int_bytes || 0;
    const totalDataBytes = wordBytes + phoneticBytes + translationBytes + intBytes;
    const totalDbBytes = result50k.bytes;
    const indexBytes = totalDbBytes - totalDataBytes;
    
    console.log('字段名'.padEnd(20) + '大小(MB)'.padStart(12) + '占比'.padStart(10));
    console.log('-'.repeat(43));
    
    const components = [
      { name: 'word 字段', bytes: wordBytes },
      { name: 'phonetic 字段', bytes: phoneticBytes },
      { name: 'translation 字段', bytes: translationBytes },
      { name: 'collins+oxford', bytes: intBytes },
      { name: '小计(字段数据)', bytes: totalDataBytes },
      { name: '索引+元数据', bytes: indexBytes },
      { name: '总计', bytes: totalDbBytes },
    ];
    
    components.forEach(comp => {
      const mb = (comp.bytes / 1024 / 1024).toFixed(2);
      const percentage = ((comp.bytes / totalDbBytes) * 100).toFixed(1);
      const name = comp.name === '小计(字段数据)' || comp.name === '总计' ? 
                   `\n${comp.name}` : comp.name;
      
      console.log(
        name.padEnd(20) +
        (mb + ' MB').padStart(12) +
        (percentage + '%').padStart(10)
      );
    });
    
    console.log('\n关键发现：');
    console.log(`- 字段数据占 ${((totalDataBytes / totalDbBytes) * 100).toFixed(1)}%`);
    console.log(`- 索引+元数据占 ${((indexBytes / totalDbBytes) * 100).toFixed(1)}%`);
    console.log(`- Translation 占字段数据的 ${((translationBytes / totalDataBytes) * 100).toFixed(1)}%\n`);
    
    testDb.close();
    compareWithOthers();
  });
}

function compareWithOthers() {
  console.log('=== 与其他词典对比 ===\n');
  
  console.log('其他人的词典: 4 MB = 50,000 词');
  console.log('计算: 4 MB / 50,000 = 81.92 字节/词\n');
  
  const result50k = results.find(r => r.words === 50000);
  if (result50k) {
    console.log(`你的数据库: ${result50k.sizeMB} MB = 50,000 词`);
    console.log(`计算: ${result50k.sizeMB} MB / 50,000 = ${result50k.bytesPerWord} 字节/词\n`);
    
    if (result50k.bytesPerWord > 82) {
      console.log('⚠️  你的数据库每词占用更多空间，可能原因：');
      console.log('1. Translation 字段更长（更详细的翻译）');
      console.log('2. 包含 phonetic 字段（其他词典可能没有）');
      console.log('3. 包含 collins 和 oxford 标记');
      console.log('4. 索引开销（SQLite 的 B-Tree 索引）\n');
      
      console.log('优化建议：');
      console.log('1. 进一步压缩 translation（限制到 100 字符）');
      console.log('2. 考虑删除 phonetic 字段（如果不需要音标）');
      console.log('3. 使用更激进的压缩算法\n');
    }
  }
  
  // 计算 30MB 可以放多少词
  console.log('=== 30MB 容量计算 ===\n');
  
  if (result50k) {
    const bytesPerWord = result50k.bytesPerWord;
    const maxWords = Math.floor(TARGET_SIZE_BYTES / bytesPerWord);
    
    console.log(`基于实测数据 (${bytesPerWord} 字节/词)：`);
    console.log(`30 MB 可容纳: ${maxWords.toLocaleString()} 词\n`);
    
    // 如果按其他人的标准
    const othersMaxWords = Math.floor(TARGET_SIZE_BYTES / 81.92);
    console.log(`如果优化到其他词典水平 (81.92 字节/词)：`);
    console.log(`30 MB 可容纳: ${othersMaxWords.toLocaleString()} 词\n`);
  }
  
  db.close();
  
  // 清理测试文件
  if (fs.existsSync(tempDbPath)) {
    fs.unlinkSync(tempDbPath);
  }
  
  console.log('测试完成！');
}

File: scripts/precise_calculation.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import sqlite3
import os
import re
import time
import random

# 路径配置
DB_PATH = os.path.join(os.path.dirname(__file__), '..', 'assets', 'ecdict-sqlite-28', 'stardict.db')
TEMP_DIR = os.path.join(os.path.dirname(__file__), '..', 'temp')

TARGET_SIZE_MB = 30
TARGET_SIZE_BYTES = TARGET_SIZE_MB * 1024 * 1024

print('=== 精确计算 30MB 可容纳词条数 ===\n')
print(f'目标大小: {TARGET_SIZE_MB} MB = {TARGET_SIZE_BYTES:,} 字节\n')

# 确保 temp 目录存在
os.makedirs(TEMP_DIR, exist_ok=True)

def compress_translation(text, max_length=150):
    """压缩翻译内容"""
    if not text:
        return ''
    
    # 删除标记
    text = re.sub(r'\[网络\][^\n]*', '', text)
    text = re.sub(r'\[地名\][^\n]*', '', text)
    text = re.sub(r'\[人名\][^\n]*', '', text)
    
    # 简化词性标记
    text = re.sub(r'\bna\.\s*', '', text)
    text = re.sub(r'\bn\.\s*', '名 ', text)
    text = re.sub(r'\bv\.\s*', '动 ', text)
    text = re.sub(r'\badj\.\s*', '形 ', text)
    text = re.sub(r'\badv\.\s*', '副 ', text)
    
    # 清理空格
    text = re.sub(r'\s+', ' ', text).strip()
    
    # 限制长度
    return text[:max_length]

def create_test_db(word_count, compression_level=150):
    """创建测试数据库"""
    # 使用唯一的文件名
    temp_db_path = os.path.join(TEMP_DIR, f'test_{word_count}_{int(time.time())}.db')
    
    # 连接源数据库
    source_conn = sqlite3.connect(DB_PATH)
    source_cursor = source_conn.cursor()
    
    # 创建测试数据库
    test_conn = sqlite3.connect(temp_db_path)
    test_cursor = test_conn.cursor()
    
    # 创建表结构
    test_cursor.execute('''
        CREATE TABLE dict (
            word TEXT PRIMARY KEY COLLATE NOCASE,
            phonetic TEXT,
            translation TEXT,
            collins INTEGER DEFAULT 0,
            oxford INTEGER DEFAULT 0
        )
    ''')
    
    # 查询数据
    source_cursor.execute(f'''
        SELECT word, phonetic, translation, collins, oxford
        FROM stardict
        WHERE oxford > 0 OR collins > 0 OR (bnc > 0 AND bnc <= 50000)
        LIMIT {word_count}
    ''')
    
    # 插入数据
    rows_inserted = 0
    for row in source_cursor.fetchall():
        word, phonetic, translation, collins, oxford = row
        
        # 压缩翻译
        translation = compress_translation(translation, compression_level)
        
        test_cursor.execute('''
            INSERT INTO dict (word, phonetic, translation, collins, oxford)
            VALUES (?, ?, ?, ?, ?)
        ''', (word, phonetic, translation, collins or 0, oxford or 0))
        
        rows_inserted += 1
    
    test_conn.commit()
    
    # VACUUM 压缩
    test_cursor.execute('VACUUM')
    test_conn.commit()
    
    # 关闭连接
    test_conn.close()
    source_conn.close()
    
    # 获取文件大小
    file_size = os.path.getsize(temp_db_path)
    
    return file_size, rows_inserted, temp_db_path

def analyze_components(word_count, db_path):
    """分析数据库成分"""
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()
    
    cursor.execute('''
        SELECT 
            SUM(LENGTH(word)) as word_bytes,
            SUM(LENGTH(COALESCE(phonetic, ''))) as phonetic_bytes,
            SUM(LENGTH(COALESCE(translation, ''))) as translation_bytes,
            COUNT(*) as count
        FROM dict
    ''')
    
    row = cursor.fetchone()
    word_bytes = row[0] or 0
    phonetic_bytes = row[1] or 0
    translation_bytes = row[2] or 0
    count = row[3] or 0
    
    # 整数字段 (collins + oxford)
    int_bytes = count * 8
    
    conn.close()
    
    return {
        'word': word_bytes,
        'phonetic': phonetic_bytes,
        'translation': translation_bytes,
        'integers': int_bytes,
        'count': count
    }

# 测试不同词条数
test_sizes = [1000, 5000, 10000, 20000, 30000, 50000]
results = []
test_db_files = []

print('正在创建测试数据库，精确测量实际大小...\n')

for test_size in test_sizes:
    print(f'测试 {test_size:,} 词条...')
    
    file_size, rows, db_path = create_test_db(test_size, compression_level=150)
    test_db_files.append(db_path)
    size_mb = file_size / 1024 / 1024
    bytes_per_word = file_size / rows
    percentage = (file_size / TARGET_SIZE_BYTES) * 100
    
    print(f'  实际大小: {size_mb:.2f} MB ({file_size:,} 字节)')
    print(f'  每词占用: {bytes_per_word:.2f} 字节')
    print(f'  占用率: {percentage:.1f}%\n')
    
    results.append({
        'words': rows,
        'bytes': file_size,
        'size_mb': size_mb,
        'bytes_per_word': bytes_per_word,
        'percentage': percentage,
        'db_path': db_path
    })

# 显示结果表格
print('=== 实际测试结果 ===\n')
print(f'{"词条数":<12} {"数据库大小":>15} {"每词占用":>12} {"占用率":>10}')
print('-' * 50)

for r in results:
    print(f'{r["words"]:<12,} {r["size_mb"]:>12.2f} MB {r["bytes_per_word"]:>9.2f} B {r["percentage"]:>8.1f}%')

print('-' * 50)
print(f'目标限制: {TARGET_SIZE_MB} MB\n')

# 分析 50000 词的成分
print('=== 数据库成分分析 ===\n')

result_50k = next((r for r in results if r['words'] == 50000), None)

if result_50k:
    components = analyze_components(50000, result_50k['db_path'])
    
    total_data_bytes = (components['word'] + components['phonetic'] + 
                       components['translation'] + components['integers'])
    total_db_bytes = result_50k['bytes']
    index_bytes = total_db_bytes - total_data_bytes
    
    print(f'基于 50,000 词的数据库 ({result_50k["size_mb"]:.2f} MB)：\n')
    print(f'{"字段名":<20} {"大小(MB)":>12} {"占比":>10}')
    print('-' * 43)
    
    items = [
        ('word 字段', components['word']),
        ('phonetic 字段', components['phonetic']),
        ('translation 字段', components['translation']),
        ('collins+oxford', components['integers']),
        ('', 0),  # 空行
        ('小计(字段数据)', total_data_bytes),
        ('索引+元数据', index_bytes),
        ('', 0),  # 空行
        ('总计', total_db_bytes),
    ]
    
    for name, byte_size in items:
        if not name:
            print()
            continue
        mb = byte_size / 1024 / 1024
        percentage = (byte_size / total_db_bytes) * 100
        print(f'{name:<20} {mb:>9.2f} MB {percentage:>8.1f}%')
    
    print('\n关键发现：')
    print(f'- 字段数据占 {(total_data_bytes / total_db_bytes) * 100:.1f}%')
    print(f'- 索引+元数据占 {(index_bytes / total_db_bytes) * 100:.1f}%')
    print(f'- Translation 占字段数据的 {(components["translation"] / total_data_bytes) * 100:.1f}%')
    print(f'- 平均 translation 长度: {components["translation"] / components["count"]:.1f} 字符\n')

# 容量计算
print('=== 容量计算 ===\n')

# 使用最大的测试结果
largest_result = results[-1]
avg_bytes = largest_result['bytes_per_word']

print(f'基于 {largest_result["words"]:,} 词的测试：')
print(f'平均每词占用: {avg_bytes:.2f} 字节\n')

# 不同压缩策略
compression_levels = [
    ('当前压缩(150字符)', 1.0),
    ('轻度压缩(200字符)', 1.15),
    ('重度压缩(100字符)', 0.85),
    ('极限压缩(50字符)', 0.70),
]

print('不同压缩策略下的最大词条数：\n')
print(f'{"压缩策略":<25} {"每词占用":>12} {"最大词条数":>15} {"实际大小":>12}')
print('-' * 65)

for name, factor in compression_levels:
    bytes_per_word = avg_bytes * factor
    max_words = int(TARGET_SIZE_BYTES / bytes_per_word)
    actual_size_mb = (max_words * bytes_per_word) / 1024 / 1024
    
    print(f'{name:<25} {bytes_per_word:>9.1f} B {max_words:>15,} {actual_size_mb:>9.2f} MB')

print()

# 与其他词典对比
print('=== 与其他词典对比 ===\n')

print('其他人的词典: 4 MB = 50,000 词')
print('计算: 4 MB / 50,000 = 81.92 字节/词\n')

largest_result = results[-1]
print(f'你的数据库: {largest_result["size_mb"]:.2f} MB = {largest_result["words"]:,} 词')
print(f'计算: {largest_result["size_mb"]:.2f} MB / {largest_result["words"]:,} = {largest_result["bytes_per_word"]:.2f} 字节/词\n')

if largest_result['bytes_per_word'] > 82:
    diff = largest_result['bytes_per_word'] - 81.92
    print(f'⚠️  你的数据库每词多占用 {diff:.2f} 字节 (+{(diff/81.92*100):.1f}%)，可能原因：')
    print('1. Translation 字段更长（更详细的翻译）')
    print('2. 包含 phonetic 字段（其他词典可能没有）')
    print('3. 包含 collins 和 oxford 标记')
    print('4. SQLite 索引开销\n')
else:
    print('✅ 你的数据库已经很优化了！\n')

# 30MB 容量计算
print('=== 30MB 容量计算 ===\n')

largest_result = results[-1]
bytes_per_word = largest_result['bytes_per_word']
max_words = int(TARGET_SIZE_BYTES / bytes_per_word)

print(f'基于实测数据 ({bytes_per_word:.2f} 字节/词)：')
print(f'30 MB 可容纳: {max_words:,} 词')
print(f'实际大小: {(max_words * bytes_per_word / 1024 / 1024):.2f} MB\n')

# 如果优化到其他词典水平
others_max_words = int(TARGET_SIZE_BYTES / 81.92)
print(f'如果优化到其他词典水平 (81.92 字节/词)：')
print(f'30 MB 可容纳: {others_max_words:,} 词')
print(f'实际大小: {(others_max_words * 81.92 / 1024 / 1024):.2f} MB\n')

# 推荐方案
print('=== 推荐方案 ===\n')

largest_result = results[-1]
# 计算不同词汇量方案
scenarios = [
    ('方案A: 最大化', 0.85, '重度压缩(100字符)'),
    ('方案B: 平衡', 1.0, '中度压缩(150字符)'),
    ('方案C: 保守', 1.15, '轻度压缩(200字符)'),
]

for name, factor, desc in scenarios:
    bytes_pw = largest_result['bytes_per_word'] * factor
    max_w = int(TARGET_SIZE_BYTES / bytes_pw)
    size = (max_w * bytes_pw) / 1024 / 1024
    
    print(f'{name}')
    print(f'  压缩策略: {desc}')
    print(f'  最大词条: {max_w:,} 词')
    print(f'  数据库大小: {size:.2f} MB')
    print(f'  剩余空间: {TARGET_SIZE_MB - size:.2f} MB')
    print()

print('测试完成！')

# 清理测试文件
for db_file in test_db_files:
    try:
        if os.path.exists(db_file):
            os.remove(db_file)
    except:
        pass
